File: Common/Utilities.Debug.vb
Imports System.Runtime.CompilerServices
Imports System.Text
Imports Media = System.Windows.Media
Imports Reflect = System.Reflection
Imports SIO = System.IO

Namespace Common

    Partial Class Utilities

        Public Class Debug

            Private Sub New()
                Throw New NotSupportedException()
            End Sub

            Public Shared Sub ShowMessageBox(text As String, Optional caption As String = "")
                Forms.MessageBox.Show(text, caption)
            End Sub

            Public Shared Function CompactStackTrace(ByVal Count As Integer) As String
                Dim ST = New StackTrace(True)
                Dim F = ST.GetFrames()

                If Count >= F.Length Then
                    Count = F.Length - 1
                End If

                Dim R = New StringBuilder()
                For I As Integer = 1 To Count
                    If I > 1 Then
                        R.Append(">"c)
                    End If
                    Dim M = F(I).GetMethod
                    R.Append(M.DeclaringType.Name) _
                     .Append("."c) _
                     .Append(M.Name) _
                     .Append(M.GetParameters().Length) _
                     .Append(":"c) _
                     .Append(F(I).GetFileLineNumber())
                Next

                Return R.ToString()
            End Function

            Public Shared Sub PrintCallInformation(ByVal ParamArray Args As Object())
                Dim R = New StringBuilder()

                Dim F = New StackFrame(1)
                Dim M = F.GetMethod()
                Dim P = M.GetParameters()
                F = New StackFrame(2, True)

                If Args.Length <> 0 And P.Length <> Args.Length Then
                    Throw New ArgumentException()
                End If

                R.Append(F.GetFileLineNumber) _
                 .Append(":") _
                 .Append(M.DeclaringType.Name) _
                 .Append("."c) _
                 .Append(M.Name) _
                 .Append("_"c) _
                 .Append(P.Length) _
                 .Append("("c)

                If Args.Length = 0 Then
                    If P.Length <> 0 Then
                        R.Append("...")
                    End If
                Else
                    Dim Bl = True
                    For I As Integer = 0 To Args.Length - 1
                        If Bl Then
                            Bl = False
                        Else
                            R.Append(", ")
                        End If
                        R.Append(P(I).Name).Append("="c).Append(Args(I))
                    Next
                End If

                R.AppendLine(")"c)

                Console.Write(R.ToString())
            End Sub

            Public Shared Function GetCallInformation(ByVal ParamArray Args As Object()) As String
                Dim R = New StringBuilder()

                Dim F = New StackFrame(1)
                Dim M = F.GetMethod()
                Dim P = M.GetParameters()
                F = New StackFrame(2, True)

                If Args.Length <> 0 And P.Length <> Args.Length Then
                    Throw New ArgumentException()
                End If

                R.Append(F.GetFileLineNumber) _
                 .Append(":") _
                 .Append(M.DeclaringType.Name) _
                 .Append("."c) _
                 .Append(M.Name) _
                 .Append(P.Length) _
                 .Append("("c)

                If Args.Length = 0 Then
                    If P.Length <> 0 Then
                        R.Append("...")
                    End If
                Else
                    Dim Bl = True
                    For Each A In Args
                        If Bl Then
                            Bl = False
                        Else
                            R.Append(", ")
                        End If
                        R.Append(A)
                    Next
                End If

                R.AppendLine(")"c)

                Return R.ToString()
            End Function

        End Class

    End Class

End Namespace

File: Common/Utilities.Debug.cs
using System;
using System.Diagnostics;
using System.Text;

namespace Ks.Common
{
    partial class Utilities
    {
        public static class Debug
        {

            public static string CompactStackTrace(int Count)
            {
                var ST = new StackTrace(true);
                var F = ST.GetFrames();

                if (Count >= F.Length)
                {
                    Count = F.Length - 1;
                }

                var R = new StringBuilder();
                for (var I = 0; I < Count; I++)
                {
                    if (I > 1)
                    {
                        R.Append('>');
                    }

                    var M = F[I].GetMethod();
                    R.Append(M.DeclaringType.Name)
                     .Append('.')
                     .Append(M.Name)
                     .Append(M.GetParameters().Length)
                     .Append(':')
                     .Append(F[I].GetFileLineNumber());
                }

                return R.ToString();
            }

            public static void PrintCallInformation(params object[] Args)
            {
                var R = new StringBuilder();

                var F = new StackFrame(1);
                var M = F.GetMethod();
                var P = M.GetParameters();
                F = new StackFrame(2, true);

                if ((Args.Length != 0) & (P.Length != Args.Length))
                {
                    throw new ArgumentException();
                }

                R.Append(F.GetFileLineNumber())
                 .Append(":")
                 .Append(M.DeclaringType.Name)
                 .Append('.')
                 .Append(M.Name)
                 .Append('_')
                 .Append(P.Length)
                 .Append('(');

                if (Args.Length == 0)
                {
                    if (P.Length != 0)
                    {
                        R.Append("...");
                    }
                }
                else
                {
                    var Bl = true;
                    for (var I = 0; I < Args.Length; I++)
                    {
                        if (Bl)
                        {
                            Bl = false;
                        }
                        else
                        {
                            R.Append(", ");
                        }

                        R.Append(P[I].Name).Append('=').Append(Args[I]);
                    }
                }

                R.Append(')').AppendLine();

                Console.Write(R.ToString());
            }

            public static string GetCallInformation(params object[] Args)
            {
                var R = new StringBuilder();

                var F = new StackFrame(1);
                var M = F.GetMethod();
                var P = M.GetParameters();
                F = new StackFrame(2, true);

                if ((Args.Length != 0) & (P.Length != Args.Length))
                {
                    throw new ArgumentException();
                }

                R.Append(F.GetFileLineNumber())
                 .Append(":")
                 .Append(M.DeclaringType.Name)
                 .Append('.')
                 .Append(M.Name)
                 .Append(P.Length)
                 .Append('(');

                if (Args.Length == 0)
                {
                    if (P.Length != 0)
                    {
                        R.Append("...");
                    }
                }
                else
                {
                    var Bl = true;
                    foreach (var A in Args)
                    {
                        if (Bl)
                        {
                            Bl = false;
                        }
                        else
                        {
                            R.Append(", ");
                        }

                        R.Append(A);
                    }
                }

                R.Append(')').AppendLine();

                return R.ToString();
            }
        }
    }
}
